File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode front = new ListNode();
        ListNode tail = front;

        // both lists have items in them
        while (list1 != null && list2 != null) {
            // if lower list val is less or equal, put that pointer to the start of other list
            if (list1.val <= list2.val) {
                tail.next = list1;
                list1 = list1.next;
            }
            else {
                tail.next = list2;
                list2 = list2.next;
            }
            tail = tail.next;
        }

        // if list1 has no items but list2 still does
        if (list1 == null && list2 != null) {
            // tail.next would go to list2
            tail.next = list2;
        } else if (list2 == null && list1 != null){
            tail.next = list1;
        }

        return front.next;
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    

    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        tail = dummy = ListNode()

        while list1 and list2:
            if list1.val < list2.val:
                tail.next = list1
                list1 = list1.next
            else:
                tail.next = list2
                list2 = list2.next
            tail = tail.next
            
        if list1 or list2:
            tail.next = list1 if list1 else list2

        return dummy.next  

    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        length = len(lists)
        if length <= 1: return None if length == 0 else lists[0]
        
        final_list = lists[0]
        for i in range(1, length):
            final_list = self.mergeTwoLists(final_list, lists[i])

        return final_list